### ISSUE TITLE
use_self lint false positive in macros

### ISSUE THREAD

**Issue by pavel-mukhanov** [rust-lang/rust-clippy#3881]
Minimal example
```rust
macro_rules! impl_binary_form_scalar {
    ($type:tt, $read:ident) => {
        impl BinaryValue for $type {
            fn to_bytes(&self) -> Vec<u8> {
                vec![*self as u8]
            }
        }
    };
}
```
This macros is invoked for 
```rust
impl_binary_form_scalar! { u8,  read_u8 }
```
and
```rust
impl_binary_form_scalar! { i8,  read_i8 }
```
Clippy suggests to replace `u8` with Self `vec![*self as Self]` which in case with `i8` is not possible.
Also, I can't figure out how to disable lint in this particular place because 
```rust
#![cfg_attr(feature="cargo-clippy", allow(clippy::use_self))]
fn to_bytes(&self) -> Vec<u8> {
```
is not working.
```
cargo +nightly clippy -V
clippy 0.0.212 (016d92d6 2019-03-10)
```

**Comment by oli-obk** [comment]
we should just silence this lint inside macros

**Comment by mati865** [comment]
:thinking: https://github.com/rust-lang/rust-clippy/issues/2098

**Comment by oli-obk** [comment]
:laughing: ok, the arguments go both ways. I guess we just need to figure out how to disable the lint with an `allow`.

@pavel-mukhanov have you tried `#[allow(clippy::use_self)]` without the cfg-attr?

**Comment by pavel-mukhanov** [comment]
@oli-obk it works if applied to `impl` defenition 👍 
```rust
#[allow(clippy::use_self)]
impl BinaryValue for $type {
     fn to_bytes(&self) -> Vec<u8> {
           vec![*self as u8]
     }
}
```

**Comment by mati865** [comment]
I was about to say you can't use `_ as Self` without triggering [E0605](https://doc.rust-lang.org/stable/error-index.html#E0605) and lint should skip not trigger in that cases.
On deeper thought it's possible to write such code: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=62c6d2c8d16f5e39702ccfc65242cbd6

**Comment by oli-obk** [comment]
Ah, we should have more fine grained lint attribute checks. We're probably using `check_item` for the `Self` and then a visitor to look into the methods. That visitor should not use `cx.span_lint` directly. Instead we might have to resort to https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/struct.TyCtxt.html#method.struct_span_lint_hir
